File: backend/runtime/internal/repository/prompt_pg.go
package repository

import (
	"context"
	"fmt"

	"gorm.io/gorm"
	"luna-ai/backend/runtime/internal/infrastructure"
)

// PromptPGRepo 提供对 prompt_templates 和 prompt_versions 表的访问
type PromptPGRepo struct {
	db *gorm.DB
}

// NewPromptPGRepo 创建 PromptPGRepo
func NewPromptPGRepo(client *infrastructure.PostgresClient) *PromptPGRepo {
	return &PromptPGRepo{
		db: client.GetDB(),
	}
}

// ListTemplates 获取所有模板列表
func (r *PromptPGRepo) ListTemplates(ctx context.Context) ([]PromptTemplate, error) {
	var templates []PromptTemplate
	if err := r.db.WithContext(ctx).Find(&templates).Error; err != nil {
		return nil, fmt.Errorf("获取模板列表失败: %w", err)
	}
	return templates, nil
}

// GetTemplate 获取模板
func (r *PromptPGRepo) GetTemplate(ctx context.Context, id string) (*PromptTemplate, error) {
	var template PromptTemplate
	if err := r.db.WithContext(ctx).First(&template, "id = ?", id).Error; err != nil {
		return nil, fmt.Errorf("获取模板失败: %w", err)
	}
	return &template, nil
}

// GetTemplateByName 获取模板
func (r *PromptPGRepo) GetTemplateByName(ctx context.Context, name string) (*PromptTemplate, error) {
	var template PromptTemplate
	if err := r.db.WithContext(ctx).First(&template, "name = ?", name).Error; err != nil {
		return nil, fmt.Errorf("获取模板失败: %w", err)
	}
	return &template, nil
}

// GetTemplatesByCategory 获取指定分类的模板
func (r *PromptPGRepo) GetTemplatesByCategory(ctx context.Context, category string) ([]PromptTemplate, error) {
	var templates []PromptTemplate
	if err := r.db.WithContext(ctx).Where("category = ?", category).Find(&templates).Error; err != nil {
		return nil, fmt.Errorf("获取分类模板失败: %w", err)
	}
	return templates, nil
}

// CreateTemplate 创建模板
func (r *PromptPGRepo) CreateTemplate(ctx context.Context, template *PromptTemplate) error {
	if err := r.db.WithContext(ctx).Create(template).Error; err != nil {
		return fmt.Errorf("创建模板失败: %w", err)
	}
	return nil
}

// UpdateTemplate 更新模板
func (r *PromptPGRepo) UpdateTemplate(ctx context.Context, template *PromptTemplate) error {
	if err := r.db.WithContext(ctx).Save(template).Error; err != nil {
		return fmt.Errorf("更新模板失败: %w", err)
	}
	return nil
}

// GetVersion 获取版本
func (r *PromptPGRepo) GetVersion(ctx context.Context, id string) (*PromptVersion, error) {
	var version PromptVersion
	if err := r.db.WithContext(ctx).First(&version, "id = ?", id).Error; err != nil {
		return nil, fmt.Errorf("获取版本失败: %w", err)
	}
	return &version, nil
}

// GetVersionsByTemplate 获取模板的所有版本
func (r *PromptPGRepo) GetVersionsByTemplate(ctx context.Context, templateID string) ([]PromptVersion, error) {
	var versions []PromptVersion
	if err := r.db.WithContext(ctx).Where("template_id = ?", templateID).Order("version_num desc").Find(&versions).Error; err != nil {
		return nil, fmt.Errorf("获取模板版本失败: %w", err)
	}
	return versions, nil
}

// CreateVersion 创建版本
func (r *PromptPGRepo) CreateVersion(ctx context.Context, version *PromptVersion) error {
	if err := r.db.WithContext(ctx).Create(version).Error; err != nil {
		return fmt.Errorf("创建版本失败: %w", err)
	}
	return nil
}

// UpdateVersion 更新版本
func (r *PromptPGRepo) UpdateVersion(ctx context.Context, version *PromptVersion) error {
	if err := r.db.WithContext(ctx).Save(version).Error; err != nil {
		return fmt.Errorf("更新版本失败: %w", err)
	}
	return nil
}

// DeleteVersion 删除版本
func (r *PromptPGRepo) DeleteVersion(ctx context.Context, id string) error {
	if err := r.db.WithContext(ctx).Delete(&PromptVersion{}, "id = ?", id).Error; err != nil {
		return fmt.Errorf("删除版本失败: %w", err)
	}
	return nil
}

// RunInTransaction 在事务中执行操作
func (r *PromptPGRepo) RunInTransaction(ctx context.Context, fn func(txRepo interface{}) error) error {
	return r.db.WithContext(ctx).Transaction(func(tx *gorm.DB) error {
		txRepo := &PromptPGRepo{db: tx}
		return fn(txRepo)
	})
}

File: backend/ai-service/app/repository/prompt_pg.py
"""
Luna AI PostgreSQL 提示词模板存储库

做什么：提供对 prompt_templates 和 prompt_versions 表的访问。
为什么这样做：管理系统提示词模板及其版本。
输入输出：
    - PromptPGRepo: 提示词模板存储库类
边界条件：
    - 支持事务操作
异常行为：
    - 数据库操作失败时抛出异常
"""

from contextlib import asynccontextmanager
from typing import Callable, Coroutine, List, Optional

from sqlalchemy import select
from sqlalchemy.ext.asyncio import AsyncSession

from app.infrastructure.postgres import PostgresClient
from app.repository.models import PromptTemplate, PromptVersion


class PromptPGRepo:
    """提供对 prompt_templates 和 prompt_versions 表的访问"""

    def __init__(self, pg_client: PostgresClient, session: Optional[AsyncSession] = None):
        self.pg_client = pg_client
        self._session = session # 用于事务中传递 session

    @asynccontextmanager
    async def _get_session(self):
        """获取 session，如果当前在事务中则返回事务 session"""
        if self._session:
            yield self._session
        else:
            async with self.pg_client.session_factory() as session:
                yield session

    async def list_templates(self) -> List[PromptTemplate]:
        """获取所有模板列表"""
        async with self._get_session() as session:
            stmt = select(PromptTemplate)
            result = await session.execute(stmt)
            return list(result.scalars().all())

    async def get_template(self, id: str) -> Optional[PromptTemplate]:
        """获取模板"""
        async with self._get_session() as session:
            stmt = select(PromptTemplate).where(PromptTemplate.id == id)
            result = await session.execute(stmt)
            return result.scalars().first()

    async def get_template_by_name(self, name: str) -> Optional[PromptTemplate]:
        """获取模板"""
        async with self._get_session() as session:
            stmt = select(PromptTemplate).where(PromptTemplate.name == name)
            result = await session.execute(stmt)
            return result.scalars().first()

    async def get_templates_by_category(self, category: str) -> List[PromptTemplate]:
        """获取指定分类的模板"""
        async with self._get_session() as session:
            stmt = select(PromptTemplate).where(PromptTemplate.category == category)
            result = await session.execute(stmt)
            return list(result.scalars().all())

    async def create_template(self, template: PromptTemplate) -> None:
        """创建模板"""
        async with self._get_session() as session:
            session.add(template)
            if not self._session:
                await session.commit()

    async def update_template(self, template: PromptTemplate) -> None:
        """更新模板"""
        async with self._get_session() as session:
            session.add(template)
            if not self._session:
                await session.commit()

    async def get_version(self, id: str) -> Optional[PromptVersion]:
        """获取版本"""
        async with self._get_session() as session:
            stmt = select(PromptVersion).where(PromptVersion.id == id)
            result = await session.execute(stmt)
            return result.scalars().first()

    async def get_versions_by_template(self, template_id: str) -> List[PromptVersion]:
        """获取模板的所有版本"""
        async with self._get_session() as session:
            stmt = (
                select(PromptVersion)
                .where(PromptVersion.template_id == template_id)
                .order_by(PromptVersion.version_num.desc())
            )
            result = await session.execute(stmt)
            return list(result.scalars().all())

    async def ensure_template_with_version(
        self,
        *,
        name: str,
        category: str,
        slot_position: str,
        content: str,
        variables: list[str],
    ) -> None:
        """
        确保指定 Prompt 模板和已发布版本存在。

        做什么：按 name 查找模板，不存在时创建模板和 published 版本。
        为什么这样做：新增 simple prompt 文件必须在服务启动时进入 PromptManager 可加载的数据源。
        输入输出：输入模板名称、分类、槽位、内容和变量列表，无返回值。
        边界条件：模板已存在时不覆盖，避免破坏用户在 Prompt 管理面板中的自定义版本。
        异常行为：数据库异常向上抛出，由启动流程记录。
        """
        async with self.pg_client.session_factory() as session:
            result = await session.execute(select(PromptTemplate).where(PromptTemplate.name == name))
            existing = result.scalars().first()
            if existing:
                return
            from app.utils.snowflake import generate_string_id

            template = PromptTemplate(
                id=generate_string_id(),
                name=name,
                category=category,
                slot_position=slot_position,
                is_system=True,
            )
            session.add(template)
            await session.flush()
            version = PromptVersion(
                id=generate_string_id(),
                template_id=template.id,
                version_num=1,
                content=content,
                variables=variables,
                status="published",
            )
            session.add(version)
            await session.flush()
            template.active_version_id = version.id
            await session.commit()

    async def create_version(self, version: PromptVersion) -> None:
        """创建版本"""
        async with self._get_session() as session:
            session.add(version)
            if not self._session:
                await session.commit()

    async def update_version(self, version: PromptVersion) -> None:
        """更新版本"""
        async with self._get_session() as session:
            session.add(version)
            if not self._session:
                await session.commit()

    async def delete_version(self, id: str) -> None:
        """删除版本"""
        async with self._get_session() as session:
            version = await session.get(PromptVersion, id)
            if version:
                await session.delete(version)
                if not self._session:
                    await session.commit()

    async def run_in_transaction(self, fn: Callable[['PromptPGRepo'], Coroutine]) -> None:
        """在事务中执行操作"""
        async with self.pg_client.session_factory() as session:
            async with session.begin():
                tx_repo = PromptPGRepo(self.pg_client, session=session)
                await fn(tx_repo)
